src/Exception/BatchException.php

```
@@ -2,61 +2,136 @@
-            }
-        }
+    /**
+     * Sets the successful response list
+     *
+     * @param BatchResult $result
+     *
+     * @internal
+     */
+    public function setResult(BatchResult $result)
+    {
+        $this->result = $result;
+    }
+
+    /**
+     * Checks if a request is successful
+     *
+     * @param RequestInterface $request
+     *
+     * @return boolean
+     */
+    public function isSuccessful(RequestInterface $request)
+    {
+        return isset($this->result) && $this->result->hasResponseFor($request);
+    }
 
-        $this->exceptions = $exceptions;
+    /**
+     * Checks if a request is failed
+     *
+     * @param RequestInterface $request
+     *
+     * @return boolean
+     */
+    public function isFailed(RequestInterface $request)
+    {
+        return $this->exceptions->contains($request);
     }
 
     /**
      * Returns all exceptions
      *
-     * @return TransferException[]
+     * @return Exception[]
      */
     public function getExceptions()
     {
-        return $this->exceptions;
+        $exceptions = [];
+
+        foreach ($this->exceptions as $request) {
+            $exceptions[] = $this->exceptions[$request];
+        }
+
+        return $exceptions;
     }
 
     /**
-     * Checks if a specific exception exists
+     * Returns an exception for a request or null if not found
      *
-     * @param TransferException $exception
+     * @param RequestInterface $request
      *
-     * @return boolean TRUE if there is the exception else FALSE.
+     * @return Exception|null
      */
-    public function hasException(TransferException $exception)
+    public function getExceptionFor(RequestInterface $request)
     {
-        return array_search($exception, $this->exceptions, true) !== false;
+        if ($this->exceptions->contains($request)) {
+            return $this->exceptions[$request];
+        }
```

lets throw an InvalidArgumentException if you request the exception for a request that is not found.